코인 리스트 화면 구현

z3rosmith opened the pull request:
* 관심/원화 탭과 정렬버튼은 아직 미완성
- changeRate에 따른 라벨 색상 변경
- filter목록을 없애고 Array의 private extension으로 filter와 sorted함수를 구현하여 코드를 간결화 시킴
- Coin타입 파일 분리
- 코인리스트 검색, 관심, 정렬할때 생기던 문제들을 해결
- 코드 간결화

z3rosmith committed:
feat: 코인리스트 검색 구현 중 #47

z3rosmith committed:
feat: 코인리스트 검색, 관심, 정렬기능 구현 완료 #47
- 코인리스트 검색, 관심, 정렬할때 생기던 문제들을 해결
- 코드 간결화

z3rosmith committed:
feat: 코인리스트 화면 구현 #47
- changeRate에 따른 라벨 색상 변경
- filter목록을 없애고 Array의 private extension으로
  filter와 sorted함수를 구현하여 코드를 간결화 시킴
- Coin타입 파일 분리

z3rosmith committed:
fix: CoinListDataManager의 delegate에 weak 적용
- CoinListDataManager의 delegate에 weak 적용

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -81,31 +141,37 @@ extension CoinListDataManager {
                                 count += 1
-                                if count == self?.coinList.count {
-                                    self?.delegate?.coinListDataManagerDidFetchCurrentPrice()
+                                if count == self?.allCoinList.count,
+                                   let self = self {
+                                    self.sortCoinList(what: .sortBoth, by: .popularity(isDescend: true), filteredBy: nil)
```
DispatchQueue Group을 활용하면 count를 사용하지 않아도 작업(코인 개수만큼 네트워크 요청)의 완료 시점을 알아낼 수 있을 것 같습니다.
``` swift
let group = DispatchGroup()

let work = DispatchWorkItem {
   // 코인 개수만큼 네트워크 요청
}
DispatchQueue.global().async(group: group, execute: work)


group.notify(queue: .main) {
  // 요청 완료 후 작업
}
```

z3rosmith replied:
설명 감사드립니다! 적용시켜보도록 할게요!

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -81,31 +141,37 @@ extension CoinListDataManager {
                 switch result {
                 case .success(let data):
                     do {
                         let response = try JSONParser().decode(data: data, type: TranscationValueObject.self)
-                        guard response.status == self?.successStatusCode else { return }
+                        guard response.status == self?.successStatusCode else {
+                            print(response.status)
+                            return
+                        }
```
`try JSONParser().decode(data: data, type: TranscationValueObject.self)` 파싱해서 데이터로 변환해주는 메서드로 따로 분리하면 코드의 뎁스를 줄일 수 있을 것 같습니다:)

z3rosmith replied:
넵 따로 분리해보도록 하겠습니다!

HJEHA commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -81,31 +141,37 @@ extension CoinListDataManager {
+        for i in 0..<allCoinList.count {
+            let api = TransactionHistoryAPI(orderCurrency: allCoinList[i].symbolName)
+            httpNetworkService.request(api: api) { [weak self] result in
                 switch result {
```
다른 branch에서 Result 타입의 익스텐션(switch를 사용하지 않고 데이터 받아오기)을 만들었는데 
머지후에 반영해도 좋을 것 같습니다:)

z3rosmith replied:
넵 머지 후에 반영하겠습니다!

HJEHA reviewed (commented):
코멘트 확인 부탁드립니다:)

donggeonoh commented on BithumbYagomAcamedy/Controller/Coin.swift:
```diff
@@ -0,0 +1,92 @@
+        guard let currentPrice = currentPrice,
+              let formattedPrice = formatter.string(for: currentPrice)
+        else {
+            return "오류발생"
```
이 오류 발생 문자열을 따로 빼는것두 나쁘지 않을 것 같기두 하네요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -7,54 +7,114 @@
+        case changeRate(isDescend: Bool)
+    }
+    
+    enum SortOption {
+        case sortFavorite
+        case sortAll
+        case sortBoth
+    }
```
SortOption은 sortFavorite = 관심, sortAll = 코인, sortBoth = 모두 인걸까요?
제생각엔 sortFavorite, sortCoin(List), sortAll 이 더 좋을 것 같습니다 ㅎㅎ

z3rosmith replied:
SortOption은 sortFavorite = 관심, sortAll = 코인, sortBoth = 모두가 맞습니다!
다만 sortCoin은 좀 애매할거같네요..

donggeonoh replied:
앗 그런가요 ㅠㅠ 알겠습니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -7,54 +7,114 @@
-                let second = $1.popularity ?? -Double.greatestFiniteMagnitude
-                return first > second
+    func toggleFavorite(coinCallingName: String, isAlreadyFavorite: Bool, filteredBy text: String?) {
+        guard let indexOfAllCoinList = allCoinList.firstIndex(where: {
+            $0.callingName == coinCallingName
+        }) else {
+            return
+        }
```
indexOfAllCoinList라고 하면 코인리스트 전체에 대한 인덱스를 가져오는 것 같아서
equalNameOfCoinList? 어떠신가요 ㅋㅋㅋ 이것도 좀 이상하네요 ㅎㅎ;

z3rosmith replied:
indexInAllCoinList는 어떨까요?

donggeonoh replied:
괜찮네요 ㅎㅎ 좋아요~

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -7,54 +7,114 @@
-                let first = $0.popularity ?? -Double.greatestFiniteMagnitude
-                let second = $1.popularity ?? -Double.greatestFiniteMagnitude
-                return first > second
+    func toggleFavorite(coinCallingName: String, isAlreadyFavorite: Bool, filteredBy text: String?) {
```
여기서 indexPath를 넘겨줄 순 없는걸까요? 아니면 눌린 셀에 대한 Coin 정보를 넘겨줄 수 있다면 그것도 하나의 방법이 될 수 있을 것 같아요.

z3rosmith replied:
표시되는 코인목록의 순서와 저장된 코인목록의 순서가 달라서 indexPath를 넘겨주는 방법으로는 어려울것같습니다.

donggeonoh replied:
그러면 코인 타입 자체를 넘겨주는 것도 고려해보시면 좋을 것 같아요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -120,50 +186,70 @@ extension CoinListDataManager {
+        let sortedCoinList: [Element]
+        switch sortType {
+        case .popularity(let isDescend):
+            if isDescend {
+                sortedCoinList = self.sorted {
+                    let first = $0.popularity ?? -Double.greatestFiniteMagnitude
+                    let second = $1.popularity ?? -Double.greatestFiniteMagnitude
+                    return first > second
+                }
```
스위프트 equatable 구현된걸 보면 매개변수 이름이 lhs, rhs로 하드라구여 (left hand side, right hand side)
first, second도 좋지만 lhs, rhs도 참고하셨으면 좋겠습니다 ㅎㅎ

z3rosmith replied:
넵 참고하겠습니다. 다만 굳이 매개변수를 나타낼 필요가 없을거같아 $0 $1을 사용하였습니다

donggeonoh replied:
아 first, second 네이밍에 대해서 말씀 드린거였습니다 ㅎㅎ

z3rosmith replied:
아 제가 잘못 읽었네요 참고하겠습니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -38,10 +41,14 @@ final class CoinListViewController: UIViewController {
         configureDataSource()
         configureCoinListController()
         coinListDataManager.fetchCoinList()
+        if let sortByPopularityButton = sortButtonStackView.subviews.first as? UIButton {
+            sortByPopularityButton.isSelected = true
+        }
```
이부분 깔끔하게 메서드화 하는 것도 좋아보여요! 균일화를 위해선?

z3rosmith replied:
넵!! 메서드화 시키겠습니다

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -56,78 +63,100 @@ final class CoinListViewController: UIViewController {
-        if sender.isSelected {
-            coinListDataManager.coinSortAction = { $0.callingName > $1.callingName }
-        } else {
-            coinListDataManager.coinSortAction = { $0.callingName < $1.callingName }
-        }
         sender.isSelected.toggle()
+        coinListDataManager.sortCoinList(
+            what: .sortAll,
+            by: .name(isDescend: sender.isSelected),
+            filteredBy: searchBar.text
+        )
+        restoreSortButtons(exclude: sender)
```
완전 깔끔하게 바뀌었네요 ㅋㅋ 전보다 훨씬 나아진 것 같습니다 👍

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -56,78 +63,100 @@ final class CoinListViewController: UIViewController {
-            }
+    // MARK: - Helper
+    
+    private func scrollToAllSectionHeader() {
```
모든 섹션 헤더를 스크롤한다..? 라고 읽혀서
scrollToCoin(List)SectionHeader() 가 어떨까요?

z3rosmith replied:
section이 favorite과 all로 나눠져있어서 scrollToCoin(List)SectionHeader()라고 하면 어느 섹션인지 애매할것같습니다!

donggeonoh replied:
섹션의 이름을 Favorite과 Coin 으로도 나누는걸 고려하는것두 좋을 것 같습니다!

z3rosmith replied:
Coin이면 어떤 코인인지 애매하지 않을까요..?

donggeonoh replied:
Coins 여도 괜찮을 것 같은데, 의미가 애매하시면 지금도 나쁘지 않을 것 같습니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -56,78 +63,100 @@ final class CoinListViewController: UIViewController {
+            ofKind: UICollectionView.elementKindSectionHeader,
+            at: IndexPath(item: 0, section: 1)
+           ) {
+            let sectionHeaderOrign = headerLayoutAttr.frame.origin
```
sectionHeaderOrign 오타나셨네요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -56,78 +63,100 @@ final class CoinListViewController: UIViewController {
-        sender.isSelected.toggle()
+    }
+    
+    private func restoreSortButtons(exclude button: UIButton) {
+        sortButtonStackView
+            .subviews
+            .map { $0 as? UIButton }
+            .compactMap { $0 }
+            .filter { $0 != button }
+            .forEach {
+                $0.isSelected = false
+            }
```
함수형 프로그래밍을 좀 해보셨나봐요... 👍

donggeonoh replied:
sortButtonStackView
            .subviews
            .map { $0 as? UIButton }
            .filter { $0 != button }
            .forEach {
                $0?.isSelected = false
            }
compactMap을 빼는 것도 괜찮을 것 같아요. 옵셔널은 equatable을 채택하고 있어서 filter 코드블럭 수행에 문제 없을 것으로 보입니다~ (forEach도 마찬가지 일 것 같아요)

z3rosmith replied:
오.. 좋은지적 감사합니다 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListViewController.swift:
```diff
@@ -56,78 +63,100 @@ final class CoinListViewController: UIViewController {
         let cellNib = UINib(nibName: "CoinListCollectionViewCell", bundle: nil)
         let coinCellRegistration = UICollectionView.CellRegistration<CoinListCollectionViewCell, Coin>(cellNib: cellNib) { cell, indexPath, item in
             cell.update(item: item)
+            cell.toggleFavorite = { [weak self] in
+                self?.coinListDataManager.toggleFavorite(
+                    coinCallingName: item.callingName,
+                    isAlreadyFavorite: item.isFavorite,
+                    filteredBy: self?.searchBar.text
+                )
+            }
```
이 때, 제생각엔 toggleFavorite 메서드에 indexPath 혹은 item 자체를 넘겨주는것이 좋을 것 같아요!

z3rosmith replied:
indexPath는 위에서 말씀드려서 넘어가겠습니다!
item을 넘겨주어도 item의 callingname밖에 사용하지 않을것같아서 굳이 그럴 필요는 없을것같습니다.

donggeonoh commented on BithumbYagomAcamedy/View/CoinListView/CoinListCollectionViewCell.swift:
```diff
@@ -9,21 +9,45 @@ import UIKit
     @IBOutlet private weak var changeRateLabel: UILabel!
     @IBOutlet private weak var changePriceLabel: UILabel!
     
+    var toggleFavorite: () -> Void = {}
```
var toggleFavorite: (() -> Void)?
로 하면 바로 초기화하지 않아도 돼서 좋을 것 같네요 ㅎㅎ

donggeonoh replied:
그리고 콜백 메서드를 실행 할 땐 toggleFavorite?()로 호출하면 될 것 같습니다~

z3rosmith replied:
좋은 지적 감사합니다.

donggeonoh reviewed (changes requested):
코멘트 단게 많네요. 천천히 답변 주셔도 좋습니다 ㅎㅎ

z3rosmith committed:
refactor: fetchCurrentPrice 함수 리팩토링 #47
- DispatchGroup을 이용해서 비동기 호출의 결과를 한번에
  알아내도록 바꾸었습니다
- parsedTranscationValueObject를 이용해서 함수의 뎁스를
  줄였습니다

donggeonoh commented on BithumbYagomAcamedy/Controller/CoinListDataManager.swift:
```diff
@@ -81,89 +141,121 @@ extension CoinListDataManager {
-        for i in 0..<coinList.count {
-            let api = TransactionHistoryAPI(orderCurrency: coinList[i].symbolName)
-            networkService.request(api: api) { [weak self] result in
+        let group = DispatchGroup()
+        for i in 0..<self.allCoinList.count {
+            group.enter()
+            let api = TransactionHistoryAPI(orderCurrency: self.allCoinList[i].symbolName)
+            self.httpNetworkService.request(api: api) { [weak self] result in
+                defer {
+                    group.leave()
+                }
```
DispatchGroup을 사용하셔서 순서를 동기화 하셨네요. 배워갑니다 👍

z3rosmith committed:
refactor: PR 논의사항 반영 #37
- PR 논의사항 반영
- Cell swipe action 추가

z3rosmith committed:
Merge branch 'develop' of https://github.com/DonggeonOh/bithumb-techcamp-ios-1st into feat/47/add-favorite-search-coinlist
